doc/reference/configuration/configuration_reference.rst

```diff
@@ -2877,6 +2878,120 @@ This section describes configuration parameters related to roles.
+        ``apply()`` should throw an error if the specified configuration can't be applied.
+
+    -   ``stop()``
+
+        Stop a role.
+        This method is called on configuration reload if the role is removed from ``roles`` for the given instance.
+        ``stop()`` should cancel all background fibers and clean up used resources.
+        If these actions can't be performed, ``stop()`` should throw an error.
+
+    -   (Optional) ``dependencies``
+
+        Define a role's dependencies as an array.
+        For roles that depend on each other, their ``validate()`` and ``apply()`` methods are executed taking into account the dependencies.
+        Suppose, there are three roles that depend on each other as follows:
+
+        ..  code-block:: none
+
+            role1
+                └─── role2
+                         └─── role3
+
+        In this case, ``validate()`` and ``apply()`` for these roles are executed in the following order:
+
+        ..  code-block:: none
+
+            role3 -> role2 -> role1
+
+
+    **Example: Built-in role**
+
+    The examples below show how to enable the router and storage-specific roles provided by the `CRUD <https://github.com/tarantool/crud>`__ module:
+
+    -   ``roles.crud-router``:
+
+        ..  literalinclude:: /code_snippets/snippets/sharding/instances.enabled/sharded_cluster_crud/config.yaml
+            :language: yaml
+            :start-at: routers
+            :end-at: roles.crud-router
+            :dedent:
+
+    -   ``roles.crud-storage``:
+
+        ..  literalinclude:: /code_snippets/snippets/sharding/instances.enabled/sharded_cluster_crud/config.yaml
+            :language: yaml
+            :start-at: storages
+            :end-at: roles.crud-storage
+            :dedent:
+
+    You can find the full example here: `sharded_cluster_crud <https://github.com/tarantool/doc/tree/latest/doc/code_snippets/snippets/sharding/instances.enabled/sharded_cluster_crud>`_.
+
+    **Example: Custom role**
+
+    The example below shows how to enable the custom ``greeter`` role for ``instance001`` and specify the configuration for this role using :ref:`roles_cfg <configuration_reference_roles_cfg>`:
+
+    ..  literalinclude:: /code_snippets/snippets/config/instances.enabled/application_role/config.yaml
+        :language: yaml
+        :start-at: instance001
+        :end-at: greeting
+        :dedent:
+
+    You can define the ``validate()``, ``apply()``, and ``stop()`` functions in the role's code as follows:
+
+    ..  literalinclude:: /code_snippets/snippets/config/instances.enabled/application_role/greeter.lua
+        :language: lua
+        :dedent:
+
+    You can find the full example here: `application_role <https://github.com/tarantool/doc/tree/latest/doc/code_snippets/snippets/config/instances.enabled/application_role>`_.
+
+    **Example: Role dependency**
+
+    The example below implements the ``greeter`` role:
```

Comment:
Just idea: how about somethings that creates an HTTP+JSON endpoint as an example? It feels much more interesting than logging to a file, while it can be written in a short way using the `http` module[^1].

We anyway lean on the `crud` module in this document, so why not?

[^1]: If you'll go this way, you may find cartridge's hack for removing routes useful. See https://github.com/tarantool/http/issues/150.